File: head_first_design_pattern/Python/observer.py
from abc import ABC, abstractmethod


# interfaces
class Subject(ABC):
    @abstractmethod
    def register_observer(self, o):
        pass

    @abstractmethod
    def remove_observer(self, o):
        pass

    @abstractmethod
    def notify_observers(self, o):
        pass


class Observer(ABC):
    @abstractmethod
    def update(self, temp, humidity, pressure):
        pass


class DisplayElement(ABC):
    @abstractmethod
    def display(self):
        pass


class WeatherDate(Subject):
    observers = []

    def __init__(self):
        self.temperature = 0
        self.humidity = 0
        self.pressure = 0
        pass

    def register_observer(self, o):
        self.observers.append(o)

    def remove_observer(self, o):
        if o in self.observers:
            self.observers.remove(o)

    def notify_observers(self):
        for x in self.observers:
            x.update(self.temperature, self.humidity, self.pressure)

    def measurements_changed(self):
        self.notify_observers()

    def set_measurements(self, temperature, humidity, pressure):
        self.temperature = temperature
        self.humidity = humidity
        self.pressure = pressure
        self.measurements_changed()


class CurrentConditionsDisplay(DisplayElement):
    def __init__(self, weather_data):
        self.temperature = 0
        self.humidity = 0
        self.pressure = 0
        self.weather = weather_data
        weather_data.register_observer(self)

    def update(self, temperature, humidity, pressure):
        self.temperature = temperature
        self.humidity = humidity
        self.pressure = pressure
        self.display()

    def display(self):
        print("Current conditions {} F degrees and {} humidity".format(self.temperature, self.humidity))


class StatisticsDisplay(DisplayElement):
    def __init__(self, weather_data):
        self.temperature = 0
        self.humidity = 0
        self.pressure = 0
        self.weather = weather_data
        weather_data.register_observer(self)

    def update(self, temperature, humidity, pressure):
        self.temperature = temperature
        self.humidity = humidity
        self.pressure = pressure
        self.display()

    def display(self):
        print("Statistics conditions {} F degrees and {} humidity".format(self.temperature, self.humidity))

weather_data = WeatherDate()
current_display = CurrentConditionsDisplay(weather_data)
statistics_display = StatisticsDisplay(weather_data)
weather_data.set_measurements(80, 65, 30.4)

File: head_first_design_pattern/C++/observer.cpp
#include<iostream>
#include<vector>

using namespace std;

// interface
class Observer {
public:
	virtual void update(float temp, float humidity, float pressure) = 0;
};

class Subject {
public:
	virtual void register_observer(Observer* o) = 0;
	virtual void remove_observer(Observer *o) = 0;
	virtual void notify_observers() = 0;
};

class DisplayElement {
public:
	virtual void display() = 0;
};


class  WeatherData :public Subject {
private:
	vector<Observer*> observers;
	float temperature;
	float humidity;
	float pressure;

public:
	WeatherData() {};
	void register_observer(Observer *o) {
		observers.push_back(o);
	};
	
	void remove_observer(Observer *o) {
		vector<Observer *> ::iterator index;
		if ((index = find(observers.begin(), observers.end(), o)) != observers.end()){
			observers.erase(index);
		}
	};

	void notify_observers() {
		for (int i = 0; i < observers.size(); i++) {
			observers[i]->update(temperature, humidity, pressure);
		}
	};

	void measurements_changed() {
		notify_observers();
	};

	void set_measurements(float temperature, float humidity, float pressure){
		this->temperature = temperature;
		this->humidity = humidity;
		this->pressure = pressure;
		measurements_changed();
	};
};

class CurrentConditionsDisplay :public Observer, DisplayElement {
private:
	float temperature;
	float humidity;
	Subject * weatherData;

public:
	CurrentConditionsDisplay(Subject * weatherData) {
		this->weatherData = weatherData;
		weatherData->register_observer(this);
	};

	void update(float temperature, float humidity, float pressure) {
		this->temperature = temperature;
		this->humidity = humidity;
		display();
	}

	void display() {
		cout <<"Current conditions:" << temperature << "F degrees and"<< humidity << endl;
	}
};

class StatisticsDisplay :public Observer, DisplayElement {
private:
	float temperature;
	float humidity;
	Subject * weatherData;

public:
	StatisticsDisplay(Subject * weatherData) {
		this->weatherData = weatherData;
		weatherData->register_observer(this);
	};

	void update(float temperature, float humidity, float pressure) {
		this->temperature = temperature;
		this->humidity = humidity;
		display();
	}

	void display() {
		cout << "StatisticsDisplay conditions:" << temperature << "F degrees and" << humidity << endl;
	}
};


int main(void)
{
	WeatherData *weather_data = new WeatherData();
	CurrentConditionsDisplay * current_display = new CurrentConditionsDisplay(weather_data);
	StatisticsDisplay* statistics_display = new StatisticsDisplay(weather_data);

	weather_data->set_measurements(80, 65, 30.4);
	while (1);

	return 0;
}